File: frontend/src/types/quiz.ts
export type QuestionType = "mcq" | "true_false" | "fill_blank" | "mixed";
export type Difficulty = "easy" | "medium" | "hard";

export interface QuizQuestion {
  id: string;
  type: string;
  difficulty: string;
  question: string;
  options: string[];
}

export interface Quiz {
  id: string;
  document_id: string;
  title: string;
  question_type: string;
  difficulty: string;
  created_at: string;
  questions: QuizQuestion[];
}

export interface QuizResultItem {
  question_id: string;
  type: string;
  question: string;
  options: string[];
  your_answer: string;
  correct_answer: string;
  is_correct: boolean;
  explanation: string;
}

export interface QuizAttemptResult {
  score: number;
  total: number;
  percentage: number;
  points: number;
  results: QuizResultItem[];
}

export interface LeaderboardEntry {
  rank: number;
  name: string;
  points: number;
  attempts: number;
  avg_percentage: number;
}

export interface QuizGenerateRequest {
  question_type: QuestionType;
  difficulty: Difficulty;
  num_questions: number;
}

File: backend/app/schemas/quiz.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class QuizGenerateRequest(BaseModel):
    question_type: str = "mcq"
    difficulty: str = "medium"
    num_questions: int = Field(default=10, ge=1, le=20)


class QuizQuestionPublic(BaseModel):
    id: str
    type: str
    difficulty: str
    question: str
    options: list[str]


class QuizResponse(BaseModel):
    id: UUID
    document_id: UUID
    title: str
    question_type: str
    difficulty: str
    created_at: datetime
    questions: list[QuizQuestionPublic]


class QuizSummary(BaseModel):
    id: UUID
    title: str
    question_type: str
    difficulty: str
    num_questions: int
    created_at: datetime


class QuizAttemptRequest(BaseModel):
    answers: dict[str, str]


class QuizResultItem(BaseModel):
    question_id: str
    type: str
    question: str
    options: list[str]
    your_answer: str
    correct_answer: str
    is_correct: bool
    explanation: str


class QuizAttemptResult(BaseModel):
    score: int
    total: int
    percentage: float
    points: int
    results: list[QuizResultItem]


class LeaderboardEntry(BaseModel):
    rank: int
    name: str
    points: int
    attempts: int
    avg_percentage: float
